### ISSUE TITLE
'threading.py(359): wait' show as 1-top hotspot in profiling table on CPU

### ISSUE THREAD

**Issue by yanbing-j** [pytorch/pytorch#188520]
### 🐛 Describe the bug

Currently, we found that the profile result of key_average table will show `threading.py(359): wait` as the first top hotspot in Xeon CPU. However, it will affect the performance analysis based on the profiling table, the total time and the hotspot are not precise.

```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  --------------------------------------------------------------------------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls                                                                      Input Shapes  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  --------------------------------------------------------------------------------  
                                threading.py(359): wait        50.00%     706.865ms        50.00%     706.865ms     706.865ms             1                                                                                []  
                       sgl_kernel::weight_packed_linear        14.49%     204.868ms        14.50%     205.043ms       6.408ms            32                                             [[1000, 2560], [18432, 2560], [], []]  
                       sgl_kernel::weight_packed_linear         8.25%     116.693ms         8.27%     116.891ms       3.653ms            32                                              [[1000, 9216], [2560, 9216], [], []]  
                       sgl_kernel::weight_packed_linear         7.62%     107.754ms         7.63%     107.883ms       4.495ms            24                                             [[1000, 2560], [12288, 2560], [], []]  
                       sgl_kernel::extend_attention_cpu         1.30%      18.422ms         1.31%      18.487ms       2.311ms             8  [[1000, 16, 256], [1000, 4, 256], [1000, 4, 256], [1000, 16, 256], [63357, 4, 25  
                           sgl_kernel::silu_and_mul_cpu         1.29%      18.288ms         1.30%      18.380ms     574.385us            32                                                                   [[1000, 18432]]  
                      sgl_kernel::causal_conv1d_fwd_cpu         0.96%      13.614ms         0.97%      13.763ms     573.454us            24         [[8192, 1000], [8192, 4], [], [1945, 8192, 3], [2], [1], [1], [], [], []]  
                    sgl_kernel::fused_rmsnorm_gated_cpu         0.65%       9.173ms         0.66%       9.300ms     387.503us            24                                           [[32000, 128], [128], [32000, 128], []]  
......
Self CPU time total: 1.312 s
```

After some investigation, I find the change of `threading.py` comes from https://github.com/pytorch/pytorch/pull/178168. I try to add https://github.com/pytorch/pytorch/pull/178168/changes#diff-857d7d889693922265a363e8ceb661dfb4d654b34d11ca669be352f45f54998dL466-L475 back, `threading.py` is gone. But I'm not sure if anything else is related.

@ryanzhang22 Could you please take a look and give some suggestions?

### Versions

PyTorch version: 2.14.0a0+gitaab577b
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: Could not collect
CMake version: version 3.27.0
Libc version: glibc-2.35

Python version: 3.12.6 | packaged by conda-forge | (main, Sep 22 2024, 14:16:49) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-124-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  384
On-line CPU(s) list:                     0-383
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) 6972P
CPU family:                              6
Model:                                   173
Thread(s) per core:                      2
Core(s) per socket:                      96
Socket(s):                               2
Stepping:                                1
CPU max MHz:                             3900.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4800.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect user_shstk avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req hfi vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr ibt amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities ibpb_exit_to_user
Virtualization:                          VT-x
L1d cache:                               9 MiB (192 instances)
L1i cache:                               12 MiB (192 instances)
L2 cache:                                384 MiB (192 instances)
L3 cache:                                960 MiB (2 instances)
NUMA node(s):                            6
NUMA node0 CPU(s):                       0-31,192-223
NUMA node1 CPU(s):                       32-63,224-255
NUMA node2 CPU(s):                       64-95,256-287
NUMA node3 CPU(s):                       96-127,288-319
NUMA node4 CPU(s):                       128-159,320-351
NUMA node5 CPU(s):                       160-191,352-383
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS Not affected; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] Could not collect
[conda] intel-cmplr-lib-ur        2024.2.0                 pypi_0    pypi
[conda] intel-openmp              2024.2.0                 pypi_0    pypi
[conda] numpy                     2.4.6                    pypi_0    pypi
[conda] optree                    0.19.1                   pypi_0    pypi
[conda] torch                     2.12.0a0+gitc51d946          pypi_0    pypi

cc @robieta @chaekit @guotuofeng @guyang3532 @dzhulgakov @davidberard98 @briancoutinho @sraikund16 @sanrise @mwootton @divyanshk @jiannanWang @scotts @ryanzhang22

**Comment by AnkitNakhawa** [comment]
Hi, I’d be interested in investigating this. I’m going to try to reproduce it with a smaller CPU-only profiler workload and determine whether Python wait events should be excluded specifically from `key_averages()` or handled another way.

Could you share the smallest workload or script that reproduces the large `threading.py(...): wait` entry? In particular, does it require SGLang’s asynchronous execution, or can it be reproduced with a standalone PyTorch workload?

**Comment by yanbing-j** [comment]
I use the following test to reproduce with PyTorch only.

```
import torch, threading, time
def worker():
    time.sleep(0.1)
t = threading.Thread(target=worker)
with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CPU], with_stack=True) as p:
    t.start()
    t.join()
avgs = p.key_averages()
threading_entries = [e for e in avgs if 'threading' in e.key]
print('torch', torch.__version__, ': threading entries =', len(threading_entries))
for e in threading_entries[:5]:
    print(' ', e.key)

```

**Comment by AnkitNakhawa** [comment]
I traced this to PR #178168 (43edd9ee5fc), which removed the C++ filter that previously excluded isPythonFunction() events from kinetoEvents. That was intentional for events() / Chrome Trace parity, but the side effect is that key_averages() now sees those events too.

I have a fix ready: add include_python_functions=False to key_averages() so Python frame events are skipped by default (with an opt-in param to restore them). Would it be possible to get the actionable label so I can open a PR?
